input/dir/dir_input_test.go

```diff
@@ -0,0 +1,103 @@
+package dir
+
+import (
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+	"time"
+
+	. "github.com/smartystreets/goconvey/convey"
+	"github.com/spacemonkeygo/errors"
+	"github.com/spacemonkeygo/errors/try"
+	"polydawn.net/repeatr/def"
+	"polydawn.net/repeatr/lib/fshash"
+	"polydawn.net/repeatr/testutil"
+)
+
+func Test(t *testing.T) {
+	Convey("Given a nonexistant path", t, func() {
+		Convey("The input config should be rejected during validation", func() {
+			correctError := false
+			try.Do(func() {
+				New(def.Input{
+					Type:     "dir",
+					Hash:     "abcd",
+					URL:      "/tmp/certainly/should/not/exist",
+					Location: "/data",
+					// REVIEW: two of these four fields were more for the executor.
+					// maybe this struct isn't the sanest to use for this arg.
+				})
+			}).Catch(def.ValidationError, func(e *errors.Error) {
+				correctError = true
+			}).Done()
+
+			So(correctError, ShouldBeTrue)
+		})
+	})
+
+	Convey("Given a directory with a mixture of files and folders", t,
+		testutil.WithTmpdir(func() {
+			pwd, _ := os.Getwd()
+			os.Mkdir("src", 0755)
+			os.Mkdir("src/a", 01777)
+			os.Mkdir("src/b", 0750)
+			f, err := os.OpenFile("src/b/c", os.O_RDWR|os.O_CREATE, 0664)
+			So(err, ShouldBeNil)
+			f.Write([]byte("zyx"))
+			So(f.Close(), ShouldBeNil)
+
+			// save attributes first because access times are conceptually insane
+			// remarkably, since the first read doesn't cause atimes to change,
+			// the inputter can capture it and we can recreate it.
+			// but that still doesn't make anything else about checking or handling it sane.
+			path0metadata := fshash.ReadMetadata("src")
+			path0metadata.Name = ""
+			path1metadata := fshash.ReadMetadata("src/a")
+			path2metadata := fshash.ReadMetadata("src/b")
+			path3metadata := fshash.ReadMetadata("src/b/c")
+
+			Convey("We can construct an input", func() {
+				inputter := New(def.Input{
+					Type: "dir",
+					Hash: "abcd",
+					URL:  filepath.Join(pwd, "src"),
+				})
+
+				Convey("Apply succeeds (hash checks pass)", func() {
+					// wait a moment before copying to decrease the odds of nanotime telling a huge, huge lie.
```

Comment:
Nanotime is *always* a lie.  Your OS can and will randomly decide to change its time reporting granularity according to its ambient temperature and how busy it feels -- even up to the order of milliseconds.  Even under "high"(ish) granularity states (read: my box at rest before I started the test), this test can totally go fast enough that all of the times writing the test files and all of the times on the files created by the copy just happen to be in the "same" nanosecond... like, yes, without this pause, the test will pass *even without* the code for forcing the fs timestamps on the copied files.